File: leetcode/0001-1000/0801-0900/0802-find-eventual-safe-states/FindEventualSafeStates.java
import java.util.ArrayList;
import java.util.List;

public class FindEventualSafeStates {
  public static void main(String[] args) {
    int[][] graph = {
        { 1, 2 },
        { 2, 3 },
        { 5 },
        { 0 },
        { 5 },
        {},
        {}
    };

    System.out.println(new Solution().eventualSafeNodes(graph));
  }
}

class Solution {
  boolean isEventual(int[][] graph, int node, boolean[] seen, boolean[] eventuals) {
    if (seen[node] || eventuals[node]) {
      return eventuals[node];
    }

    seen[node] = true;

    for (int neighbour : graph[node]) {
      if (!isEventual(graph, neighbour, seen, eventuals)) {
        return false;
      }
    }

    eventuals[node] = true;

    return true;
  }

  public List<Integer> eventualSafeNodes(int[][] graph) {
    // Number of nodes
    int n = graph.length;

    boolean[] eventuals = new boolean[n];
    boolean[] seen = new boolean[n];

    for (int i = 0; i < n; i++) {
      if (isEventual(graph, i, seen, eventuals)) {
        eventuals[i] = true;
      }
    }

    List<Integer> results = new ArrayList<>();

    for (int i = 0; i < n; i++) {
      if (eventuals[i]) {
        results.add(i);
      }
    }
    return results;
  }
}

File: leetcode/0001-1000/0801-0900/0801-0810/0802-find-eventual-safe-states/FindEventualSafeStates.ts
function eventualSafeNodes(graph: number[][]): number[] {
  const n = graph.length;
  const eventuals = new Array(n).fill(false);
  const visited = new Array(n).fill(false);


  const isEventual = (i: number): boolean => {
    if (visited[i]) {
      return eventuals[i];
    }

    visited[i] = true;

    for (const neighbour of graph[i]) {
      if (!isEventual(neighbour)) {
        return false;
      }
    }

    eventuals[i] = true;

    return eventuals[i];
  }

  for (let i = 0; i < n; i++) {
    isEventual(i);
  }

  return eventuals.map((v, i) => v ? i : null).filter(v => v !== null);
};

console.log(eventualSafeNodes([[1, 2], [2, 3], [5], [0], [5], [], []]));
